File: plugins/blackbox/osv.py
"""Client-side OSV vulnerability lookup for dependency auto-discovery.

A tiny, dependency-free helper around ``https://api.osv.dev/v1/query``. It is
the DISCOVERY nomination layer for dependencies: when an install is detected
whose package is NOT already in the graph ruleset, :func:`lookup` asks OSV
whether that exact ``package@version`` is known-vulnerable. If so, the caller
auto-submits a *candidate* dependency threat.

Design constraints (all enforced here):

* **stdlib only** — ``urllib``; no new third-party dependency.
* **fail-open** — any transport/parse error returns ``None`` (no finding).
* **short timeout** — never delays the agent loop meaningfully.
* **privacy** — only OSV-*vulnerable* installs are ever surfaced; a clean
  package returns ``None`` and is never reported.
* **cached** — an in-memory dedupe cache keyed by ``eco:name@version`` so a
  repeated install in the same process makes at most one network call.
"""

from __future__ import annotations

import json
import logging
import threading
import urllib.error
import urllib.request
from typing import Any, Dict, Optional

logger = logging.getLogger(__name__)

_OSV_URL = "https://api.osv.dev/v1/query"
_TIMEOUT = 3.0

#: Blackbox ecosystem slug → OSV ecosystem name. ``homebrew`` has no OSV
#: ecosystem, so it is intentionally absent (skipped, never looked up).
_ECOSYSTEM_MAP = {
    "npm": "npm",
    "pypi": "PyPI",
    "cargo": "crates.io",
    "rubygems": "RubyGems",
}

# In-memory result cache. Value is the finding dict or None (clean/skip).
_cache: Dict[str, Optional[Dict[str, str]]] = {}
_cache_lock = threading.Lock()


def osv_ecosystem(ecosystem: str) -> Optional[str]:
    """Map a Blackbox ecosystem slug to its OSV name, or ``None`` to skip."""
    return _ECOSYSTEM_MAP.get((ecosystem or "").strip().lower())


def _severity_of(vuln: Dict[str, Any]) -> str:
    """Best-effort severity from an OSV vuln record (defaults to ``high``)."""
    # OSV database_specific.severity is the most common human label.
    dbs = vuln.get("database_specific")
    if isinstance(dbs, dict):
        raw = str(dbs.get("severity") or "").strip().lower()
        if raw:
            return raw
    # Fall back to ecosystem-specific severity blocks when present.
    for aff in vuln.get("affected") or []:
        if isinstance(aff, dict):
            aff_dbs = aff.get("database_specific")
            if isinstance(aff_dbs, dict):
                raw = str(aff_dbs.get("severity") or "").strip().lower()
                if raw:
                    return raw
    return "high"


def _query(osv_eco: str, name: str, version: str) -> Optional[Dict[str, Any]]:
    body = json.dumps({"package": {"ecosystem": osv_eco, "name": name}, "version": version}).encode("utf-8")
    req = urllib.request.Request(
        _OSV_URL, data=body, headers={"Content-Type": "application/json", "Accept": "application/json"}, method="POST"
    )
    try:
        with urllib.request.urlopen(req, timeout=_TIMEOUT) as resp:
            raw = resp.read()
    except (urllib.error.URLError, urllib.error.HTTPError, OSError, TimeoutError) as exc:
        logger.debug("blackbox: OSV query failed for %s@%s: %s", name, version, exc)
        return None
    try:
        return json.loads(raw.decode("utf-8"))
    except Exception:  # pragma: no cover - defensive
        return None


def lookup(ecosystem: str, name: str, version: str) -> Optional[Dict[str, str]]:
    """Return ``{advisory_id, severity}`` if OSV knows *name@version* vulnerable.

    Returns ``None`` when the package is clean, the ecosystem is unsupported,
    the version is missing, or anything fails (fail-open). Cached per process.
    """
    eco = (ecosystem or "").strip().lower()
    name = (name or "").strip()
    version = (version or "").strip()
    if not name or not version:
        return None
    osv_eco = osv_ecosystem(eco)
    if not osv_eco:
        return None
    key = f"{eco}:{name.lower()}@{version}"
    with _cache_lock:
        if key in _cache:
            return _cache[key]
    result: Optional[Dict[str, str]] = None
    data = _query(osv_eco, name, version)
    if isinstance(data, dict):
        vulns = data.get("vulns")
        if isinstance(vulns, list) and vulns:
            first = vulns[0] if isinstance(vulns[0], dict) else {}
            advisory_id = str(first.get("id") or "OSV")
            result = {"advisory_id": advisory_id, "severity": _severity_of(first)}
    with _cache_lock:
        _cache[key] = result
    return result

File: integrations/openclaw/src/osv.ts
/**
 * Client-side OSV vulnerability lookup for dependency auto-discovery.
 *
 * A tiny, dependency-free helper around `https://api.osv.dev/v1/query` using the
 * global `fetch` (Node >=18). It is the DISCOVERY nomination layer for
 * dependencies: when an install is detected whose package is NOT already in the
 * graph ruleset, `lookup` asks OSV whether that exact `package@version` is
 * known-vulnerable. If so, the caller auto-submits a candidate dependency threat.
 *
 * Faithful port of `plugins/blackbox/osv.py`. Design constraints:
 *   - global `fetch` only — no new runtime dependency.
 *   - fail-open — any transport/parse error resolves to null (no finding).
 *   - short timeout — never delays the agent loop meaningfully.
 *   - privacy — only OSV-*vulnerable* installs are ever surfaced; a clean
 *     package resolves to null and is never reported.
 *   - cached — an in-memory dedupe cache keyed by `eco:name@version` so a
 *     repeated install in the same process makes at most one network call.
 */
import type { OsvHit } from "./detection.js";
import type { BlackboxSeverity } from "./quads.js";

const OSV_URL = "https://api.osv.dev/v1/query";
const TIMEOUT_MS = 3000;

/**
 * Blackbox ecosystem slug → OSV ecosystem name. `homebrew` has no OSV
 * ecosystem, so it is intentionally absent (skipped, never looked up).
 */
const ECOSYSTEM_MAP: Record<string, string> = {
  npm: "npm",
  pypi: "PyPI",
  cargo: "crates.io",
  rubygems: "RubyGems",
};

// In-memory result cache. Value is the OsvHit or null (clean/skip).
const cache = new Map<string, OsvHit | null>();

/** Map a Blackbox ecosystem slug to its OSV name, or null to skip. */
export function osvEcosystem(ecosystem: string): string | null {
  return ECOSYSTEM_MAP[(ecosystem || "").trim().toLowerCase()] ?? null;
}

/** Best-effort severity from an OSV vuln record (defaults to `high`). */
function severityOf(vuln: Record<string, unknown>): BlackboxSeverity {
  const dbs = vuln.database_specific;
  if (dbs && typeof dbs === "object") {
    const raw = String((dbs as Record<string, unknown>).severity ?? "").trim().toLowerCase();
    if (raw) return raw as BlackboxSeverity;
  }
  const affected = Array.isArray(vuln.affected) ? vuln.affected : [];
  for (const aff of affected) {
    if (aff && typeof aff === "object") {
      const affDbs = (aff as Record<string, unknown>).database_specific;
      if (affDbs && typeof affDbs === "object") {
        const raw = String((affDbs as Record<string, unknown>).severity ?? "").trim().toLowerCase();
        if (raw) return raw as BlackboxSeverity;
      }
    }
  }
  return "high";
}

async function query(
  osvEco: string,
  name: string,
  version: string,
): Promise<Record<string, unknown> | null> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    const res = await fetch(OSV_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "application/json" },
      body: JSON.stringify({ package: { ecosystem: osvEco, name }, version }),
      signal: controller.signal,
    });
    if (!res.ok) return null;
    return (await res.json()) as Record<string, unknown>;
  } catch {
    return null; // fail open
  } finally {
    clearTimeout(timer);
  }
}

/**
 * Return `{advisoryId, severity}` if OSV knows `name@version` vulnerable.
 * Resolves to null when the package is clean, the ecosystem is unsupported, the
 * version is missing, or anything fails (fail-open). Cached per process.
 * Port of Python `osv.lookup`.
 */
export async function lookup(
  ecosystem: string,
  name: string,
  version: string,
): Promise<OsvHit | null> {
  const eco = (ecosystem || "").trim().toLowerCase();
  const pkg = (name || "").trim();
  const ver = (version || "").trim();
  if (!pkg || !ver) return null;
  const osvEco = osvEcosystem(eco);
  if (!osvEco) return null;
  const key = `${eco}:${pkg.toLowerCase()}@${ver}`;
  if (cache.has(key)) return cache.get(key) ?? null;
  let result: OsvHit | null = null;
  const data = await query(osvEco, pkg, ver);
  if (data && typeof data === "object") {
    const vulns = (data as Record<string, unknown>).vulns;
    if (Array.isArray(vulns) && vulns.length > 0) {
      const first = (vulns[0] && typeof vulns[0] === "object" ? vulns[0] : {}) as Record<
        string,
        unknown
      >;
      const advisoryId = String(first.id ?? "OSV");
      result = { advisoryId, severity: severityOf(first) };
    }
  }
  cache.set(key, result);
  return result;
}
